vis/optimizer.py

```diff
@@ -8,63 +8,71 @@
-
-
-def _identity(x):
-    return x
-
-
 class Optimizer(object):
 
-    def __init__(self, input_tensor, losses, input_range=(0, 255), wrt_tensor=None, norm_grads=True):
+    def __init__(self, input_tensors, losses, input_ranges=(0, 255), wrt_tensors=None, norm_grads=True):
         """Creates an optimizer that minimizes weighted loss function.
 
         Args:
-            input_tensor: An input tensor of shape: `(samples, channels, image_dims...)` if `image_data_format=
-                channels_first` or `(samples, image_dims..., channels)` if `image_data_format=channels_last`.
+            input_tensors: An input tensor or list of input tensor.
+                The shape of an input tensor is `(samples, channels, image_dims...)` if `image_data_format=
+                channels_first`, Or it's `(samples, image_dims..., channels)` if `image_data_format=channels_last`.
             losses: List of ([Loss](vis.losses#Loss), weight) tuples.
-            input_range: Specifies the input range as a `(min, max)` tuple. This is used to rescale the
+            input_range: A tuple of length 2 or list of tuple of length 2.
+                Specifies the input range as a `(min, max)` tuple. This is used to rescale the
                 final optimized input to the given range. (Default value=(0, 255))
-            wrt_tensor: Short for, with respect to. This instructs the optimizer that the aggregate loss from `losses`
-                should be minimized with respect to `wrt_tensor`.
-                `wrt_tensor` can be any tensor that is part of the model graph. Default value is set to None
-                which means that loss will simply be minimized with respect to `input_tensor`.
+                This option is one tuple if `input_tensors` is a tensor or all the tensors are the same shape.
+            wrt_tensors: A tensor or list of tensor.
+                Short for, with respect to. This instructs the optimizer that the aggregate loss from `losses`
+                should be minimized with respect to `wrt_tensors`.
+                `wrt_tensors` can be any tensor that is part of the model graph. Default value is set to None
+                which means that loss will simply be minimized with respect to `input_tensors`.
             norm_grads: True to normalize gradients. Normalization avoids very small or large gradients and ensures
                 a smooth gradient gradient descent process. If you want the actual gradient
                 (for example, visualizing attention), set this to false.
         """
-        self.input_tensor = input_tensor
-        self.input_range = input_range
+        self.input_tensors = utils.listify(input_tensors)
+        self.input_ranges = utils.listify(input_ranges)
+        if len(self.input_tensors) != len(self.input_ranges):
+            if len(self.input_ranges) == 1:
+                for i in range(1, len(self.input_tensors)):
+                    self.input_ranges.append(self.input_ranges[0])
```

Comment:
mention this in the doc. "If a single range is provided it will be assumed for all input_tensors".